frontend/main.js

```diff
@@ -136,13 +136,40 @@ async function startApp () {
         this.ephemeral.finishedLogin = 'no'
         router.currentRoute.path !== '/' && router.push({ path: '/' }).catch(console.error)
       })
-
       sbp('okTurtles.data/set', 'BANNER', this.$refs.bannerGeneral)
       // call from anywhere in the app:
       // sbp('okTurtles.data/get', 'BANNER').show(L('Trying to reconnect...'), 'wifi')
       // sbp('okTurtles.data/get', 'BANNER').danger(L('message'), 'icon-type')
       // sbp('okTurtles.data/get', 'BANNER').clean()
+      sbp('okTurtles.data/apply', PUBSUB_INSTANCE, (instance) => {
+        const banner = this.$refs.bannerGeneral
+        // Allow to access `L` inside event handlers.
+        const L = this.L.bind(this)
 
+        Object.assign(instance.customEventHandlers, {
+          offline () {
+            banner.show(L('Your device appears to be offline.'), 'wifi')
+          },
+          online () {
+            banner.clean()
+          },
+          'reconnection-attempt' () {
+            banner.show(L('Trying to reconnect...'), 'wifi')
+          },
+          'reconnection-failed' () {
+            banner.show(L('Cannot connect to the server.'), 'wifi')
+          },
+          'reconnection-succeeded' () {
+            banner.clean()
+          }
+        })
+      })
+      // Useful in case the app is started in offline mode.
```

Comment:
So this is where I think we come to a situation that we didn't fully address in the previous PR. Correct me if I'm wrong, but AFAICT assigning to `customEventHandlers` like this will override any other handlers. And it's not even clear to me whether assigning to them after instance creation will work..

Can you modify the way that events work in the pubsub client so that this doesn't happen? One suggestion I have is to make full use of `okTurtles.events` for all of the event handling, since it doesn't have this problem. It would involve modifying the way events are handled in `shared/pubsub.js`.

Reply:
> And it's not even clear to me whether assigning to them after instance creation will work..

OK, I verified that it works, but I'm not sure why.

```js
  for (const name of Object.keys(defaultClientEventHandlers)) {
    client.listeners[name] = (event) => {
      const customHandler = client.customEventHandlers[name]
```

This code is from `createClient`, and I assume this is the only place where `customEventHandlers` is used. Initially it is `null`, so don't `const customHandler` get assigned to `null`? Why is it that assigning `customEventHandlers` still works after calling `createClient`? I must be missing something.

Reply:
> Correct me if I'm wrong, but AFAICT assigning to customEventHandlers like this will override any other handlers

It's true that it will override a previously assigned *custom* handler if there is one.

However, it will never override any so-called *default* event handler since they are hardcoded in `shared/pubsub.js` and cannot be mutated from outside that file.

Reply:
Alright, we can go with this. It works well enough.

I also deleted/retract my second question, as I see now why it works. (I must be having an off day).

Reply:
Well I have to admit that this code was not the easiest to follow. Dynamic event programming is a bit tricky! 😄